### ISSUE TITLE
Bluetooth mac-address not showing in settings

### ISSUE THREAD

**Issue by taaem** [kimmoli/sfos-onyx-issues#8]
The Bluetooth Mac isn't showing up in settings

**Comment by taaem** [comment]
May be related to #1

**Comment by kimmoli** [comment]
Comment from qml

```
// aboutSettings.bluetoothAddress may be 00:00:00:00:00 if the adapter could not
// be initialized at start-up, use BluetoothInfo instead (guarantees a valid address)
```

And, well, it uses BluetoothInfo

**Comment by kimmoli** [comment]
Available NV data, where
447 = Bluetooth Device Address, System
4678 = WLAN MAC Address, WLAN
6855 = OEM ITEM 3, Factory

```
[root@localhost oemnvitems]# xxd 447
0000000: 1234 59fb eec0                           ..Y...
[root@localhost oemnvitems]# xxd 4678
0000000: 1234 59fb eec0                           ..Y...
```

Dump says it is identical with wlan MAC address.. afaiu this is possible

**Comment by kimmoli** [comment]
hcitool dev says 

```
hci0    00:00:00:00:5A:AD
```

Found this

```
/data/property/persist.service.bdroid.bdaddr
22:22:10:e6:a9:18
```

that is `[persist.service.bdroid.bdaddr]: [22:22:10:e6:a9:18]`

**Comment by kimmoli** [comment]
Commented out stuff from init.qcom-common.rc

```
on property:vold.decrypt=trigger_restart_framework
    #start config_bluetooth
    start wcnss-service

#service config_bluetooth /system/bin/sh /system/etc/init.qcom.bt.sh "onboot"
#    class core
#    user root
#    group bluetooth net_bt_admin qcom_diag radio
#    oneshot

#service hciattach /system/bin/sh /system/etc/init.qcom.bt.sh
#    class late_start
#    user bluetooth
#    group bluetooth net_bt_admin qcom_diag radio
#    disabled
#    oneshot

#on property:bluetooth.hciattach=true
#    start hciattach
```

and let bluetooth be started from systemd-service

**Comment by kimmoli** [comment]
In Jolla (sbj) address is in NV here `xxd /dev/block/platform/msm_sdcc.1/by-name/QOTP | head`
and it is written by `/usr/lib/oneshot.d/bdaddr-from-factory-data` to file `/var/lib/bluetooth/board-address`

**Comment by kimmoli** [comment]
disabled also `/usr/bin/droid/droid-hcismd-up.sh`
then `echo 1 > /sys/module/hci_smd/parameters/hcismd_set`
and right after it `/sys/class/bluetooth/hci0/address` is all zeros `00:00:00:00:00:00`
but after `rfkill unblock bluetooth` it goes to `00:00:00:00:5A:AD`

**Comment by kimmoli** [comment]
The address in settins comes from `/var/lib/bluetooth/board-address`
Tha can be set with oneshot from `persist.service.bdroid.bdaddr` property (?)
But how to take this address really in use ???

**Comment by kimmoli** [comment]
http://forum.xda-developers.com/showpost.php?p=61774700&postcount=43
says 

> I have also found what breaks the Bluetooth address — it should be configured by /system/bin/hci_qcomm_init, but that binary is linked with libbtnv.so, which is missing. You need to pull /system/vendor/lib/libbtnv.so from the stock ROM. After adding this file Bluetooth seems to work 

This file is in place, `/system/vendor/lib/libbtnv.so`

```
echo 0 > /sys/module/hci_smd/parameters/hcismd_set
setprop ro.qualcomm.bt.hci_transport smd
# initialise chip
BLUETOOTH_MAC=`cat /var/lib/bluetooth/board-address`
/system/bin/hci_qcomm_init -d /dev/ttysmd3 -s 3000000 -i 115200 -r 32000000 -p 0 -P 1 -b $BLUETOOTH_MAC
echo 1 > /sys/module/hci_smd/parameters/hcismd_set
sleep 1
rfkill unblock bluetooth
sleep 1
hciconfig hci0 up
hcitool lescan
```

**Comment by kimmoli** [comment]
droid-hcismd-up.service - start later

```
[Unit]
Description=Enable Bluetooth HCI over SMD
DefaultDependencies=false
After=local-fs.target bluetooth.service

[Service]
Type=oneshot
ExecStart=/bin/sh /usr/bin/droid/droid-hcismd-up.sh
RemainAfterExit=yes

[Install]
WantedBy=multi-user.target
```

/usr/bin/droid/droid-hcismd-up.sh - edited script

```
#!/bin/sh

# transport is smd
setprop ro.qualcomm.bt.hci_transport smd

# initialise chip and board-address with correct mac
BLUETOOTH_MAC=`/system/bin/btnvtool -p 2>&1| sed -n 's/ //g;s/.*://;s/\./\:/g;y/abcdef/ABCDEF/;1p'`
echo Setting bluetooth address to $BLUETOOTH_MAC | systemd-cat -p info -t "droid-hcismd-up.sh"

echo $BLUETOOTH_MAC > /var/lib/bluetooth/board-address
/system/bin/hci_qcomm_init -d /dev/ttysmd3 -s 3000000 -i 115200 -r 32000000 -p 0 -P 1 -b $BLUETOOTH_MAC

# Maximum number of attempts to enable hcismd to try to get
# hci0 to come online.  Writing to sysfs too early seems to
# not work, so we loop.
MAXTRIES=15

seq 1 $MAXTRIES | while read i ; do
    echo 1 > /sys/module/hci_smd/parameters/hcismd_set
    if [ -e /sys/class/bluetooth/hci0 ] ; then
        # found hci0, exit successfully
        exit 0
    fi
    sleep 1
    if [ $i == $MAXTRIES ] ; then
        # must have gotten through all our retries, fail
        exit 1
    fi
done
```

/lib/udev/rules.d/10-rfkill.rules - udev rule to trigger on rfkill bluetooth

```
SUBSYSTEM=="rfkill", ATTR{type}=="bluetooth", RUN+="/usr/bin/droid/bluetooth-rfkill.sh"
```

/usr/bin/droid/bluetooth-rfkill.sh - new script to perform lescan only on first UI interaction

```
#!/bin/bash

if [ ${RFKILL_STATE} = 2 -o ${RFKILL_STATE} = 0 ]; then
    echo "bluetooth stopped" | systemd-cat -p info -t "bluetooth-rfkill.sh"
else
    echo "bluetooth started" | systemd-cat -p info -t "bluetooth-rfkill.sh"
    sleep 1
    if [ "`getprop droid.bt.scanned`" = "true" ]; then
      echo "lescan already done" | systemd-cat -p info -t "bluetooth-rfkill.sh"
      exit 0
    fi
    echo "performing initial lescan" | systemd-cat -p info -t "bluetooth-rfkill.sh"
    hcitool lescan &
    pid=$!
    sleep 1
    kill -INT $pid
    if [ "`getprop droid.bt.scanned`" = "false" ]; then
      setprop droid.bt.scanned true
    else
      setprop droid.bt.scanned false
    fi
    echo "initial lescan done" | systemd-cat -p info -t "bluetooth-rfkill.sh"
fi
```

**Comment by kimmoli** [comment]
also spotted this fro, journal

```
May 07 12:27:59 Jolla jolla-rfkill-hciwait[782]: Not waiting for HCI events, ID='onyx' not sbj
```

**Comment by kimmoli** [comment]
https://github.com/kimmoli/droid-config-onyx/commit/1d78022705e7b19186f70d49098994ba7c1d7cad
https://github.com/kimmoli/android_device_oneplus_onyx/commit/ca2ca4e131a39c0a7ff27ed375073af3e236a6db
